Cabal/Distribution/FooBar.hs

```diff
@@ -0,0 +1,56 @@
+module Distribution.FooBar where
+
+import           Data.Maybe
+import qualified Data.Map as Map
+
+import           Distribution.Package
+import           Distribution.PackageDescription
+
+-- | Desugar a "build-tools" entry into proper a executable dependency if
+-- possible.
+--
+-- An entry can be so desguared in two cases:
+--
+-- 1. The name in build-tools matches a locally defined executable.  The
+--    executable dependency produced is on that exe in the current package.
+--
+-- 2. The name in build-tools matches a hard-coded set of known tools.  For now,
+--    the executable dependency produced is one an executable in a package of
+--    the same, but the hard-coding could just as well be per-key.
+--
+-- The first cases matches first.
+desugarBuildTool :: PackageDescription
+                 -> LegacyExeDependency
+                 -> Maybe ExeDependency
+desugarBuildTool pkg led =
+  if foundLocal
+  then Just $ ExeDependency (packageName pkg) toolName reqVer
+  else Map.lookup name whiteMap
+  where
+    LegacyExeDependency name reqVer = led
+    toolName = mkUnqualComponentName name
+    foundLocal = toolName `elem` map exeName (executables pkg)
+    whitelist = [ "hscolour", "haddock", "happy", "alex", "hsc2hs", "c2hs"
+                , "cpphs", "greencard"]
+    whiteMap  = Map.fromList $ flip map whitelist $ \n ->
+      (n, ExeDependency (mkPackageName n) (mkUnqualComponentName n) reqVer)
+
+-- | Get everything from "tool-depends", along with entries from "build-tools"
+-- that we know how to desugar.
+--
+-- This should almost always be used instead of just accessing the
+-- `toolDepends` field directly.
```

Comment:
I added some longer documentation to the relevant fields in the top commit. I'll also grep at the end to make sure nothing is lurking.